frontend/src/metabase/containers/MoveModal.tsx

```diff
@@ -1,69 +1,110 @@
-import CollectionPicker from "metabase/containers/CollectionPicker";
-import Button from "metabase/core/components/Button";
-import type { Collection, CollectionId } from "metabase-types/api";
-
-import { ButtonContainer } from "./MoveModal.styled";
+import { isItemCollection } from "metabase/collections/utils";
+import {
+  CollectionPickerModal,
+  type CollectionPickerItem,
+} from "metabase/common/components/EntityPicker";
+import type { CollectionId, CollectionItem } from "metabase-types/api";
 
 interface MoveModalProps {
   title: string;
   onClose: () => void;
-  onMove: (collection: any) => void;
-  initialCollectionId?: number | string | null;
+  onMove: (item: { id: CollectionId }) => void;
+  initialCollectionId: CollectionId;
+  movingCollectionId?: CollectionId;
 }
 
 export const MoveModal = ({
   title,
   onClose,
   onMove,
   initialCollectionId,
+  movingCollectionId,
 }: MoveModalProps) => {
-  const [selectedCollectionId, setSelectedCollectionId] =
-    useState(initialCollectionId);
+  // if we are moving a collection, we can't move it into itself or any of its children
+  const shouldDisableItem = movingCollectionId
+    ? (item: CollectionPickerItem) =>
+        Boolean(
+          item.id === movingCollectionId ||
+            item?.location?.split("/").includes(String(movingCollectionId)),
+        )
+    : undefined;
+
+  return (
+    <CollectionPickerModal
+      title={title}
+      value={{
+        id: initialCollectionId,
+        model: "collection",
+      }}
+      onChange={newCollection => onMove({ id: newCollection.id })}
+      options={{
+        showSearch: true,
+        allowCreateNew: true,
+        hasConfirmButtons: true,
+        showRootCollection: true,
+        showPersonalCollections: true,
+        confirmButtonText: t`Move`,
+      }}
+      shouldDisableItem={shouldDisableItem}
+      onClose={onClose}
+    />
+  );
+};
+
+interface BulkMoveModalProps {
+  onClose: () => void;
+  onMove: (item: { id: CollectionId }) => void;
+  selectedItems: CollectionItem[];
+  initialCollectionId: CollectionId;
+}
+
+export const BulkMoveModal = ({
+  onClose,
+  onMove,
+  selectedItems,
+  initialCollectionId,
+}: BulkMoveModalProps) => {
+  const movingCollectionIds = selectedItems
+    .filter((item: CollectionItem) => isItemCollection(item))
+    .map((item: CollectionItem) => String(item.id));
 
-  const [creatingCollection, setCreatingCollection] = useState(false);
-  const [openCollectionId, setOpenCollectionId] = useState<CollectionId>();
+  // if the move set includes collections, we can't move into any of them
+  const shouldDisableItem = movingCollectionIds.length
+    ? (item: CollectionPickerItem) => {
+        const collectionItemFullPath =
+          item?.location?.split("/").map(String).concat(String(item.id)) ?? [];
+        return (
+          _.intersection(collectionItemFullPath, movingCollectionIds).length > 0
+        );
+      }
+    : undefined;
```

Comment:
this is also new functionality, if any of the moving items are collections, we disable them in the UI so we don't trick people into thinking they can move things there.